File: src/main/java/me/dyzjct/kura/module/modules/combat/AutoTotem.java
package me.dyzjct.kura.module.modules.combat;

import me.dyzjct.kura.module.Category;
import me.dyzjct.kura.module.Module;
import me.dyzjct.kura.setting.Setting;
import me.dyzjct.kura.utils.mc.EntityUtil;
import net.minecraft.client.Minecraft;
import net.minecraft.client.gui.inventory.GuiContainer;
import net.minecraft.client.gui.inventory.GuiInventory;
import net.minecraft.init.Items;
import net.minecraft.inventory.ClickType;
import net.minecraft.item.ItemStack;

import java.util.HashMap;
import java.util.Map;
import java.util.concurrent.atomic.AtomicInteger;

@Module.Info(name = "AutoTotem", description = "null", category = Category.COMBAT)
public class AutoTotem extends Module {
    public static AutoTotem INSTANCE = new AutoTotem();
    public static Minecraft mc = Minecraft.getMinecraft();
    public Setting<Boolean> soft = bsetting("Soft", false);
    public Setting<Boolean> strict = bsetting("Strict", false);
    public int numOfTotems;
    public int preferredTotemSlot;

    public static Map<Integer, ItemStack> getInventoryAndHotbarSlots() {
        return getInventorySlots(9);
    }

    public static Map<Integer, ItemStack> getInventorySlots(int current) {
        Map<Integer, ItemStack> fullInventorySlots = new HashMap<>();
        while (current <= 44) {
            fullInventorySlots.put(current, mc.player.inventoryContainer.getInventory().get(current));
            current++;
        }
        return fullInventorySlots;
    }

    @Override
    public void onUpdate() {

        if (mc.player == null) {
            return;
        }
        if (!findTotems()) {
            return;
        }
        if (mc.currentScreen instanceof GuiContainer && !(mc.currentScreen instanceof GuiInventory)) {
            return;
        }
        if (mc.currentScreen instanceof GuiInventory) {
            return;
        }

        if (soft.getValue()) {
            if (mc.player.getHeldItemOffhand().getItem().equals(Items.AIR)) {
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                }
                mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player);
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                    return;
                }
                mc.playerController.windowClick(0, 45, 0, ClickType.PICKUP, mc.player);
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                    return;
                }
                mc.playerController.updateController();
            }

        } else {

            if (!mc.player.getHeldItemOffhand().getItem().equals(Items.TOTEM_OF_UNDYING)) {
                boolean offhandEmptyPreSwitch = mc.player.getHeldItemOffhand().getItem().equals(Items.AIR);
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                }
                mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player);
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                    return;
                }
                mc.playerController.windowClick(0, 45, 0, ClickType.PICKUP, mc.player);
                if (strict.getValue() && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0, 0, 0);
                    return;
                }
                if (!offhandEmptyPreSwitch) {
                    mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player);
                }
                mc.playerController.updateController();

            }

        }

    }

    public boolean findTotems() {
        this.numOfTotems = 0;
        AtomicInteger preferredTotemSlotStackSize = new AtomicInteger();
        preferredTotemSlotStackSize.set(Integer.MIN_VALUE);
        getInventoryAndHotbarSlots().forEach((slotKey, slotValue) -> {
            int numOfTotemsInStack = 0;
            if (slotValue.getItem().equals(Items.TOTEM_OF_UNDYING)) {
                numOfTotemsInStack = slotValue.getCount();
                if (preferredTotemSlotStackSize.get() < numOfTotemsInStack) {
                    preferredTotemSlotStackSize.set(numOfTotemsInStack);
                    preferredTotemSlot = slotKey;
                }
            }

            this.numOfTotems = this.numOfTotems + numOfTotemsInStack;

        });
        if (mc.player.getHeldItemOffhand().getItem().equals(Items.TOTEM_OF_UNDYING)) {
            this.numOfTotems = this.numOfTotems + mc.player.getHeldItemOffhand().getCount();
        }
        return this.numOfTotems != 0;
    }

    @Override
    public String getHudInfo() {
        return "[" + this.numOfTotems + "]";
    }
}

File: src/main/java/me/windyteam/kura/module/modules/combat/AutoTotem.kt
package me.windyteam.kura.module.modules.combat

import me.windyteam.kura.module.Category
import me.windyteam.kura.module.Module
import me.windyteam.kura.setting.Setting
import me.windyteam.kura.utils.mc.EntityUtil
import net.minecraft.client.Minecraft
import net.minecraft.client.gui.inventory.GuiContainer
import net.minecraft.client.gui.inventory.GuiInventory
import net.minecraft.init.Items
import net.minecraft.inventory.ClickType
import net.minecraft.item.ItemStack
import net.minecraft.network.play.client.CPacketPlayer
import java.util.concurrent.atomic.AtomicInteger

@Module.Info(name = "AutoTotem", description = "null", category = Category.COMBAT)
object AutoTotem : Module() {
    var soft: Setting<Boolean> = bsetting("Soft", false)
    private var strict: Setting<Boolean> = bsetting("Strict", false)
    private var numOfTotems = 0
    private var preferredTotemSlot = 0
    override fun onUpdate() {
        if (mc.player == null) {
            return
        }
        if (!findTotems()) {
            return
        }
        if (mc.currentScreen is GuiContainer && mc.currentScreen !is GuiInventory) {
            return
        }
        if (mc.currentScreen is GuiInventory) {
            return
        }
        if (soft.value) {
            if (mc.player.heldItemOffhand.getItem() == Items.AIR) {
                if (strict.value){
                    Module.mc.player.connection.sendPacket(
                        CPacketPlayer.Position(
                            Module.mc.player.posX,
                            Module.mc.player.posY - 10,
                            Module.mc.player.posZ,
                            false
                        )
                    )
                }
                if (strict.value && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0.0, 0.0, 0.0)
                }
                mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player)
                mc.playerController.windowClick(0, 45, 0, ClickType.PICKUP, mc.player)
                mc.playerController.updateController()
                if (strict.value){
                    Module.mc.player.connection.sendPacket(
                        CPacketPlayer.Position(
                            Module.mc.player.posX,
                            Module.mc.player.posY + 0,
                            Module.mc.player.posZ,
                            false
                        )
                    )
                }
            }
        } else {
            if (mc.player.heldItemOffhand.getItem() != Items.TOTEM_OF_UNDYING) {
                val offhandEmptyPreSwitch = mc.player.heldItemOffhand.getItem() == Items.AIR
                if (strict.value){
                    Module.mc.player.connection.sendPacket(
                        CPacketPlayer.Position(
                            Module.mc.player.posX,
                            Module.mc.player.posY - 10,
                            Module.mc.player.posZ,
                            false
                        )
                    )
                }
                if (strict.value && EntityUtil.isMoving()) {
                    mc.player.setVelocity(0.0, 0.0, 0.0)
                }
                mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player)
                mc.playerController.windowClick(0, 45, 0, ClickType.PICKUP, mc.player)
                if (!offhandEmptyPreSwitch) {
                    mc.playerController.windowClick(0, preferredTotemSlot, 0, ClickType.PICKUP, mc.player)
                }
                mc.playerController.updateController()
                if (strict.value){
                    Module.mc.player.connection.sendPacket(
                        CPacketPlayer.Position(
                            Module.mc.player.posX,
                            Module.mc.player.posY + 0,
                            Module.mc.player.posZ,
                            false
                        )
                    )
                }
            }
        }
    }

    private fun findTotems(): Boolean {
        numOfTotems = 0
        val preferredTotemSlotStackSize = AtomicInteger()
        preferredTotemSlotStackSize.set(Int.MIN_VALUE)
        inventoryAndHotbarSlots.forEach { (slotKey: Int, slotValue: ItemStack) ->
            var numOfTotemsInStack = 0
            if (slotValue.getItem() == Items.TOTEM_OF_UNDYING) {
                numOfTotemsInStack = slotValue.count
                if (preferredTotemSlotStackSize.get() < numOfTotemsInStack) {
                    preferredTotemSlotStackSize.set(numOfTotemsInStack)
                    preferredTotemSlot = slotKey
                }
            }
            numOfTotems += numOfTotemsInStack
        }
        if (mc.player.heldItemOffhand.getItem() == Items.TOTEM_OF_UNDYING) {
            numOfTotems += mc.player.heldItemOffhand.count
        }
        return numOfTotems != 0
    }

    override fun getHudInfo(): String {
        return "[$numOfTotems]"
    }

    private var mc = Minecraft.getMinecraft()
    private val inventoryAndHotbarSlots: Map<Int, ItemStack>
        get() = getInventorySlots(9)

    private fun getInventorySlots(current: Int): Map<Int, ItemStack> {
        var current = current
        val fullInventorySlots: MutableMap<Int, ItemStack> = HashMap()
        while (current <= 44) {
            fullInventorySlots[current] = mc.player.inventoryContainer.inventory[current]
            current++
        }
        return fullInventorySlots
    }
}